### PR TITLE
C++: support for designated initializers

### PR DESCRIPTION
**Pull request by nickrolfe**
I've added new tuples to associate aggregate literals, their initializer expressions, and the field or element index being initialized.

This will need to be merged in sync with the internal extractor PR.

This PR also includes the comment-only dbscheme changes from #18, to avoid an extra upgrade script.

### CODE REVIEW

**Comment by rdmarsh2** [comment]
Can we get a test with out-of-order designated initializers? Something like ```someStruct {.j = 1, .i = 2}```. I'd also like to see a test of how this interacts with child ordering; before this PR, designated struct initializers led to creating two children for each initializer.

**Comment by nickrolfe** [comment]
@rdmarsh2 for your first request, I think the initializers on lines 63-64 of `aggregate_literals.c` show the right output:

```
| aggregate_literals.c:62:13:65:7 | {...} | aggregate_literals.c:6:8:6:22 | someOtherStruct | aggregate_literals.c:7:9:7:9 | a | aggregate_literals.c:64:14:64:16 | ... / ... |
| aggregate_literals.c:62:13:65:7 | {...} | aggregate_literals.c:6:8:6:22 | someOtherStruct | aggregate_literals.c:8:9:8:9 | b | aggregate_literals.c:63:14:63:16 | ... * ... |
```

For your request about child ordering, can  you expand on what sort of query you'd like to see? (I think I can explain the problem you saw, and it should now be fixed, but I'm happy to improve the test).

**Comment by rdmarsh2** [comment]
Those initializers look right; missed it on my first read through.

The queries I was observing issues with were:
```from ClassAggregateLiteral cal, int i
select cal, cal.getType().getUnspecifiedType(), i, cal.getChild(i)```
and
```from ClassAggregateLiteral cal, int i, Field f
where cal.getFieldExpr(f) = cal.getChild(i)
select cal, cal.getType().getUnspecifiedType(), i, cal.getChild(i), f```

**Comment by jbj** [comment]
@dave-bartolomeo can you have a look to ensure that the extracted data is suitable for IR translation?

**Comment by nickrolfe** [comment]
I've added some tests to show that the child expressions of an aggregate literal are in the correct order (and match the correct fields/array elements).

File: cpp/ql/src/semmle/code/cpp/exprs/Literal.qll [hunk 1]
```diff
@@ -173,7 +173,7 @@ class ClassAggregateLiteral extends AggregateLiteral {
    */
   Expr getFieldExpr(Field field) {
     field = classType.getAField() and
-    result = getChild(field.getInitializationOrder())
+    aggregate_field_init(this, result, field)
```
Comment (ian-semmle) [inline cpp/ql/src/semmle/code/cpp/exprs/Literal.qll]: This should be `aggregate_field_init(underlyingElement(this), unresolveElement(result), unresolveElement(field))`

File: cpp/ql/src/semmle/code/cpp/exprs/Literal.qll [hunk 2]
```diff
@@ -230,7 +230,7 @@ class ArrayAggregateLiteral extends AggregateLiteral {
    * element `elementIndex`, if present.
    */
   Expr getElementExpr(int elementIndex) {
-    result = getChild(elementIndex)
+    aggregate_array_init(this, result, elementIndex)
```
Comment (ian-semmle) [inline cpp/ql/src/semmle/code/cpp/exprs/Literal.qll]: Similarly here

File: cpp/ql/src/semmlecode.cpp.dbscheme [hunk 3]
```diff
@@ -1400,6 +1441,18 @@ new_array_allocated_type(
...
+    int initializer: @expr ref,
+    int field: @membervariable ref
+);
+
+aggregate_array_init(
+    unique int aggregate: @aggregateliteral ref,
+    int initializer: @expr ref,
+    int element_index: int ref
+);
+
```
Comment (ian-semmle) [inline cpp/ql/src/semmlecode.cpp.dbscheme]: Could you add qldoc for these please?

**Comment by ian-semmle** [comment]
We already know the field declaration order, but the source code order is extra information that is potentially useful.

**Comment by nickrolfe** [comment]
> the order of the children in the `AggregateLiteral` will match the order of the init expression in the source code, not field declaration order, right?

Right.

I was going to make the same point as @ian.

Also, changing the order of the child exprs would be a largely orthogonal extractor change to the ones made to support designated initializers. I'm sure we could do it, but I'd like to get this PR merged first (especially since the 1.18 deadline is looming).

**Comment by dave-bartolomeo** [comment]
LGTM

FWIW, constructor init lists seem to show up in field declaration order, regardless of init list order.

The IR currently ignores child order, and sorts initializers by field declaration order, so this change shouldn't impact the IR. Even if it did, the IR is easier to change after feature freeze than the extractor.

I assume I shouldn't approve this until the extractor change goes into the private repo?

**Comment by nickrolfe** [comment]
Approval is harmless, but it should only be merged at the same time as the private PR.

BTW, should I add a release note?

**Comment by jbj** [comment]
Yes, definitely add a release note for this change and also other user-visible extractor features added for 1.18. #24 probably qualifies and maybe also the support for `_FloatN` and `_FloatNx`.

**Pull request merged**